File: src/py_robot_nav/py_robot_nav/laserscan_interpolator.py
import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import LaserScan
from scipy.interpolate import interp1d
import numpy as np


class Interpolator(Node):
    def __init__(self):
        super().__init__("laserscan_interpolator")

        # Declare parameters
        self.declare_parameter("input_topic", "/scan")
        self.declare_parameter("output_topic", "/interpolated_scan")
        self.declare_parameter("target_angle_increment", 0.5)
        self.declare_parameter("distance_threshold", 0.5)

        input_topic = self.get_parameter("input_topic").value
        output_topic = self.get_parameter("output_topic").value
        self.target_angle_increment = self.get_parameter("target_angle_increment").value
        self.distance_threshold = self.get_parameter("distance_threshold").value

        self.subscription = self.create_subscription(
            LaserScan,
            input_topic,
            self.callback,
            qos_profile=qos_profile_sensor_data,
        )

        self.publisher = self.create_publisher(
            LaserScan,
            output_topic,
            qos_profile=qos_profile_sensor_data,
        )

    def resample(
        self, source_list, original_angle_step, target_angle_step, distance_threshold
    ):
        """
        Resample laser scan data with intelligent handling of inf/nan values.

        Args:
            source_list: Original laser scan ranges
            original_angle_step: Angular resolution of source data (degrees)
            target_angle_step: Desired angular resolution (degrees)
            distance_threshold: Max Euclidean distance to keep points in same subrange (meters)

        Returns:
            Resampled laser scan with uniform angular resolution
        """
        source_array = np.array(source_list)
        interpolation_factor = original_angle_step / target_angle_step
        target_length = int(round((2 * np.pi) / target_angle_step))

        # Step 1: Find valid points (not inf, not nan)
        valid_mask = np.isfinite(source_array)
        valid_indices = np.where(valid_mask)[0]
        valid_values = source_array[valid_mask]

        if len(valid_values) < 2:
            return [np.inf] * target_length

        # Step 2: Split into subranges based on:
        #   a) Angular continuity (consecutive indices)
        #   b) Spatial continuity (Euclidean distance threshold)
        subranges = []
        current_start_idx = valid_indices[0]
        current_indices = [valid_indices[0]]
        current_values = [valid_values[0]]

        for i in range(1, len(valid_values)):
            # Check if there's an angular gap (non-consecutive indices)
            angular_gap = (valid_indices[i] - valid_indices[i - 1]) > 1

            # Calculate actual Euclidean distance between points
            # Using law of cosines: d² = r₁² + r₂² - 2*r₁*r₂*cos(θ)
            r1 = valid_values[i - 1]
            r2 = valid_values[i]
            angle_diff_deg = (
                valid_indices[i] - valid_indices[i - 1]
            ) * original_angle_step
            angle_diff_rad = np.deg2rad(angle_diff_deg)

            euclidean_distance = np.sqrt(
                r1**2 + r2**2 - 2 * r1 * r2 * np.cos(angle_diff_rad)
            )

            spatial_discontinuity = euclidean_distance > distance_threshold

            if spatial_discontinuity:
                # Save current subrange
                subranges.append(
                    {
                        "start_idx": current_start_idx,
                        "end_idx": valid_indices[i - 1],  # Store end index too
                        "indices": current_indices.copy(),
                        "values": current_values.copy(),
                    }
                )
                # Start new subrange
                current_start_idx = valid_indices[i]
                current_indices = [valid_indices[i]]
                current_values = [valid_values[i]]
            else:
                # Continue current subrange
                current_indices.append(valid_indices[i])
                current_values.append(valid_values[i])

        # Add the last subrange
        subranges.append(
            {
                "start_idx": current_start_idx,
                "end_idx": valid_indices[-1],
                "indices": current_indices,
                "values": current_values,
            }
        )

        # Step 3: Interpolate each subrange independently
        result = np.full(target_length, np.inf)

        for subrange in subranges:
            values = np.array(subrange["values"])
            original_length = len(values)

            # Calculate target positions for this subrange
            # Start and end positions in the target array
            start_pos = int(round(subrange["start_idx"] * interpolation_factor))
            end_pos = int(round(subrange["end_idx"] * interpolation_factor))

            # Ensure we stay within bounds
            start_pos = max(0, min(start_pos, target_length - 1))
            end_pos = max(0, min(end_pos, target_length - 1))

            if start_pos >= end_pos:
                # Single point or degenerate case
                if 0 <= start_pos < target_length:
                    result[start_pos] = values[0] if len(values) > 0 else np.inf
                continue

            # Calculate exact number of points needed to densely fill this range
            new_length = end_pos - start_pos + 1

            if original_length < 2:
                # Single point - fill entire range with same value
                result[start_pos : end_pos + 1] = values[0]
            else:
                # Interpolate to fill exactly the required positions
                x_original = np.linspace(0, 1, original_length)
                x_new = np.linspace(0, 1, new_length)

                interpolator = interp1d(
                    x_original, values, kind="linear", fill_value="extrapolate"
                )
                interpolated_values = interpolator(x_new)

                # Fill the range completely (inclusive of both start and end)
                result[start_pos : end_pos + 1] = interpolated_values

        return result.tolist()

    def callback(self, msg: LaserScan):
        if not msg.ranges or len(msg.ranges) < 2:
            self.publisher.publish(msg)
            return

        current_inc = msg.angle_increment
        target_inc = self.target_angle_increment

        new_ranges = self.resample(
            msg.ranges,
            current_inc,
            target_inc,
            self.distance_threshold,
        )
        new_intensities = None
        if msg.intensities:
            new_intensities = self.resample(
                msg.intensities,
                current_inc,
                target_inc,
                self.distance_threshold,
            )

        new_inc = (msg.angle_max - msg.angle_min) / (len(new_ranges) - 1)
        # print(f"{current_inc=} {target_inc=} {interp_factor=} {new_inc=}")

        interp_factor = current_inc / target_inc

        new_msg = LaserScan()
        new_msg.header = msg.header
        new_msg.angle_min = msg.angle_min
        new_msg.angle_max = msg.angle_max
        new_msg.angle_increment = new_inc
        new_msg.time_increment = msg.time_increment / interp_factor
        new_msg.scan_time = msg.scan_time
        new_msg.range_min = msg.range_min
        new_msg.range_max = msg.range_max
        new_msg.ranges = new_ranges
        if new_intensities:
            new_msg.intensities = new_intensities

        # print("received", len(msg.ranges), "sent", len(new_msg.ranges))
        # print(new_msg.ranges)
        self.publisher.publish(new_msg)


def main(args=None):
    rclpy.init(args=args)
    node = Interpolator()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/agx_navigation/pointcloud_utils/src/laserscan_interpolator.cpp
#include "pointcloud_utils/laserscan_interpolator.hpp"
#include <algorithm>
#include <cmath>
#include <rclcpp_components/register_node_macro.hpp>

namespace pointcloud_utils {

LaserScanInterpolator::LaserScanInterpolator(const rclcpp::NodeOptions &options)
    : Node("laserscan_interpolator", options) {
  this->declare_parameter<std::string>("input_topic", "/scan");
  this->declare_parameter<std::string>("output_topic", "/interpolated_scan");
  this->declare_parameter<float>(
      "target_angle_increment",
      std::round((2.0f * static_cast<float>(M_PI)) / 360.0f)); // radians
  this->declare_parameter<float>("distance_threshold", 0.5f);  // metres

  const auto input_topic = this->get_parameter("input_topic").as_string();
  const auto output_topic = this->get_parameter("output_topic").as_string();
  target_angle_increment_ = static_cast<float>(
      this->get_parameter("target_angle_increment").as_double());
  distance_threshold_ =
      static_cast<float>(this->get_parameter("distance_threshold").as_double());

  // Use sensor-data QoS (best-effort, volatile) - matches typical lidar drivers
  const auto qos = rclcpp::SensorDataQoS();

  // Subscribe with a unique_ptr callback signature so the middleware can
  // hand us ownership of the message without copying.
  subscription_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
      input_topic, qos,
      std::bind(&LaserScanInterpolator::scan_callback, this,
                std::placeholders::_1));

  publisher_ =
      this->create_publisher<sensor_msgs::msg::LaserScan>(output_topic, qos);

  RCLCPP_INFO(this->get_logger(),
              "Interpolator ready  [in: %s → out: %s | target_inc: %.4f rad | "
              "dist_thresh: %.2f m]",
              input_topic.c_str(), output_topic.c_str(),
              target_angle_increment_, distance_threshold_);
}

void LaserScanInterpolator::scan_callback(
    sensor_msgs::msg::LaserScan::UniquePtr msg) {
  if (msg->ranges.size() < 2) {
    publisher_->publish(std::move(msg));
    return;
  }

  const float current_inc = msg->angle_increment;
  const float target_inc = target_angle_increment_;
  const float interp_factor = current_inc / target_inc;

  // Resample ranges
  auto new_ranges =
      resample(msg->ranges, current_inc, target_inc, distance_threshold_);

  // Resample intensities if present
  std::vector<float> new_intensities;
  if (!msg->intensities.empty()) {
    new_intensities = resample(msg->intensities, current_inc, target_inc,
                               distance_threshold_);
  }

  // Build output message via unique_ptr for zero-copy publish
  auto out = std::make_unique<sensor_msgs::msg::LaserScan>();

  out->header = msg->header;
  out->angle_min = msg->angle_min;
  out->angle_max = msg->angle_max;
  out->angle_increment = (new_ranges.size() > 1)
                             ? (msg->angle_max - msg->angle_min) /
                                   static_cast<float>(new_ranges.size() - 1)
                             : msg->angle_increment;
  out->time_increment = msg->time_increment / interp_factor;
  out->scan_time = msg->scan_time;
  out->range_min = msg->range_min;
  out->range_max = msg->range_max;
  out->ranges = std::move(new_ranges);

  if (!new_intensities.empty()) {
    out->intensities = std::move(new_intensities);
  }

  publisher_->publish(std::move(out));
}

std::vector<float> LaserScanInterpolator::resample(
    const std::vector<float> &source, float original_angle_step,
    float target_angle_step, float distance_threshold) const {
  const float interp_factor = original_angle_step / target_angle_step;
  const int target_length = static_cast<int>(
      std::round((2.0f * static_cast<float>(M_PI)) / target_angle_step));

  // Collect valid (finite) points
  std::vector<int> valid_indices;
  std::vector<float> valid_values;
  valid_indices.reserve(source.size());
  valid_values.reserve(source.size());

  for (size_t i = 0; i < source.size(); ++i) {
    if (std::isfinite(source[i])) {
      valid_indices.push_back(static_cast<int>(i));
      valid_values.push_back(source[i]);
    }
  }

  if (valid_values.size() < 2) {
    return std::vector<float>(target_length,
                              std::numeric_limits<float>::infinity());
  }

  // Split into subranges on spatial discontinuities
  std::vector<Subrange> subranges;
  Subrange current;
  current.start_idx = valid_indices[0];
  current.indices = {valid_indices[0]};
  current.values = {valid_values[0]};

  for (size_t i = 1; i < valid_values.size(); ++i) {
    const float r1 = valid_values[i - 1];
    const float r2 = valid_values[i];

    // Angle gap in radians between the two valid points
    const float angle_diff =
        static_cast<float>(valid_indices[i] - valid_indices[i - 1]) *
        original_angle_step;

    // Law of cosines:  d^2 = r1^2 + r2^2 − 2*r1*r2*cos(delta theta)
    const float euclidean_dist =
        std::sqrt(r1 * r1 + r2 * r2 - 2.0f * r1 * r2 * std::cos(angle_diff));

    if (euclidean_dist > distance_threshold) {
      // Close off current subrange, start a new one
      current.end_idx = valid_indices[i - 1];
      subranges.push_back(std::move(current));

      current = Subrange{};
      current.start_idx = valid_indices[i];
      current.indices = {valid_indices[i]};
      current.values = {valid_values[i]};
    } else {
      current.indices.push_back(valid_indices[i]);
      current.values.push_back(valid_values[i]);
    }
  }

  // Finalize the last subrange
  current.end_idx = valid_indices.back();
  subranges.push_back(std::move(current));

  // Interpolate each subrange independently
  std::vector<float> result(target_length,
                            std::numeric_limits<float>::infinity());

  for (const auto &sr : subranges) {
    const int original_len = static_cast<int>(sr.values.size());

    // Map original index range → target index range
    int start_pos = static_cast<int>(std::round(sr.start_idx * interp_factor));
    int end_pos = static_cast<int>(std::round(sr.end_idx * interp_factor));

    start_pos = std::clamp(start_pos, 0, target_length - 1);
    end_pos = std::clamp(end_pos, 0, target_length - 1);

    if (start_pos >= end_pos) {
      // Degenerate / single-point subrange
      if (start_pos >= 0 && start_pos < target_length && !sr.values.empty()) {
        result[start_pos] = sr.values.front();
      }
      continue;
    }

    const int new_len = end_pos - start_pos + 1;

    if (original_len < 2) {
      // Single source point -> fill entire target span with that value
      std::fill(result.begin() + start_pos, result.begin() + end_pos + 1,
                sr.values.front());
      continue;
    }

    // Linear interpolation (equivalent to scipy interp1d "linear")
    // Map both original and new indices onto [0, 1] and lerp.
    for (int j = 0; j < new_len; ++j) {
      // Position in [0, 1] within this subrange
      const float t = static_cast<float>(j) / static_cast<float>(new_len - 1);

      // Which segment of the original data does t fall into?
      const float src_pos = t * static_cast<float>(original_len - 1);
      const int lo = static_cast<int>(std::floor(src_pos));
      const int hi = std::min(lo + 1, original_len - 1);
      const float frac = src_pos - static_cast<float>(lo);

      result[start_pos + j] =
          sr.values[lo] * (1.0f - frac) + sr.values[hi] * frac;
    }
  }

  return result;
}

} // namespace pointcloud_utils

// Register as a composable node so it can be loaded into a component container
RCLCPP_COMPONENTS_REGISTER_NODE(pointcloud_utils::LaserScanInterpolator)
